# TV Script Generation

In this project, you'll generate your own [Seinfeld](https://en.wikipedia.org/wiki/Seinfeld) TV scripts using RNNs.  You'll be using part of the [Seinfeld dataset](https://www.kaggle.com/thec03u5/seinfeld-chronicles#scripts.csv) of scripts from 9 seasons.  The Neural Network you'll build will generate a new ,"fake" TV script, based on patterns it recognizes in this training data.

## Get the Data

The data is already provided for you in '/content/drive/MyDrive/Colab Notebooks/data/Seinfeld_Scripts.txt' and you're encouraged to open that file and look at the text. 
>* As a first step, we'll load in this data and look at some samples. 
* Then, you'll be tasked with defining and training an RNN to generate a new script!

In [1]:
# Add the local_modules directory to the set of paths
# Python uses to look for imports.
import sys
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [2]:
sys.path.append('/content/drive/MyDrive/Colab')


In [3]:
%cd /content/drive/MyDrive/Colab
!ls -l

/content/drive/MyDrive/Colab
total 86071
-rw------- 1 root root    62980 Nov 22 14:50 Character_Level_RNN_Solution.ipynb
drwx------ 2 root root     4096 Dec  1 08:16 data
-rw------- 1 root root    43854 Dec  4 17:16 dlnd_tv_script_generation.ipynb
-rw------- 1 root root     1352 Dec  4 14:35 generated_script_1.txt
-rw------- 1 root root  2563587 Dec  4 15:32 preprocess.p
-rw------- 1 root root    10307 Sep  9 21:10 problem_unittests.py
drwx------ 2 root root     4096 Dec  1 21:11 __pycache__
-rw------- 1 root root 85442883 Dec  4 14:35 trained_rnn.pt
-rw------- 1 root root     1407 Dec  1 20:55 udacity_helper.py


In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# load in data
import udacity_helper as helper
data_dir = '/content/drive/MyDrive/Colab/data/Seinfeld_Scripts.txt'
text = helper.load_data(data_dir)

## Explore the Data
Play around with `view_line_range` to view different parts of the data. This will give you a sense of the data you'll be working with. You can see, for example, that it is all lowercase text, and each new line of dialogue is separated by a newline character `\n`.

In [5]:
view_line_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in text.split()})))

lines = text.split('\n')
print('Number of lines: {}'.format(len(lines)))
word_count_line = [len(line.split()) for line in lines]
print('Average number of words in each line: {}'.format(np.average(word_count_line)))

print()
print('The lines {} to {}:'.format(*view_line_range))
print('\n'.join(text.split('\n')[view_line_range[0]:view_line_range[1]]))

Dataset Stats
Roughly the number of unique words: 46367
Number of lines: 109233
Average number of words in each line: 5.544240293684143

The lines 0 to 10:
jerry: do you know what this is all about? do you know, why were here? to be out, this is out...and out is one of the single most enjoyable experiences of life. people...did you ever hear people talking about we should go out? this is what theyre talking about...this whole thing, were all out now, no one is home. not one person here is home, were all out! there are people trying to find us, they dont know where we are. (on an imaginary phone) did you ring?, i cant find him. where did he go? he didnt tell me where he was going. he must have gone out. you wanna go out you get ready, you pick out the clothes, right? you take the shower, you get all ready, get the cash, get your friends, the car, the spot, the reservation...then youre standing around, what do you do? you go we gotta be getting back. once youre out, you wanna get back! y

---
## Implement Pre-processing Functions
The first thing to do to any dataset is pre-processing.  Implement the following pre-processing functions below:
- Lookup Table
- Tokenize Punctuation

### Lookup Table
To create a word embedding, you first need to transform the words to ids.  In this function, create two dictionaries:
- Dictionary to go from the words to an id, we'll call `vocab_to_int`
- Dictionary to go from the id to word, we'll call `int_to_vocab`

Return these dictionaries in the following **tuple** `(vocab_to_int, int_to_vocab)`

In [6]:
import problem_unittests as tests
from collections import Counter

def create_lookup_tables(text):
    """
    Create lookup tables for vocabulary
    :param text: The text of tv scripts split into words
    :return: A tuple of dicts (vocab_to_int, int_to_vocab)
    """
    # DONE: Implement Function
    #Start by getting the frequency of each of the words in the text
    word_counts = Counter(text)
        
    sorted_words = sorted(word_counts, key=word_counts.get, reverse=True)
    
    # create int_to_vocab dictionaries
    int_to_vocab = {ii: word for ii, word in enumerate(sorted_words)}
    vocab_to_int = {word: ii for ii, word in int_to_vocab.items()}    
    
    # return tuple
    return (vocab_to_int, int_to_vocab)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_create_lookup_tables(create_lookup_tables)

Tests Passed


### Tokenize Punctuation
We'll be splitting the script into a word array using spaces as delimiters.  However, punctuations like periods and exclamation marks can create multiple ids for the same word. For example, "bye" and "bye!" would generate two different word ids.

Implement the function `token_lookup` to return a dict that will be used to tokenize symbols like "!" into "||Exclamation_Mark||".  Create a dictionary for the following symbols where the symbol is the key and value is the token:
- Period ( **.** )
- Comma ( **,** )
- Quotation Mark ( **"** )
- Semicolon ( **;** )
- Exclamation mark ( **!** )
- Question mark ( **?** )
- Left Parentheses ( **(** )
- Right Parentheses ( **)** )
- Dash ( **-** )
- Return ( **\n** )

This dictionary will be used to tokenize the symbols and add the delimiter (space) around it.  This separates each symbols as its own word, making it easier for the neural network to predict the next word. Make sure you don't use a value that could be confused as a word; for example, instead of using the value "dash", try using something like "||dash||".

In [7]:
def token_lookup():
    """
    Generate a dict to turn punctuation into a token.
    :return: Tokenized dictionary where the key is the punctuation and the value is the token
    """
    # DONE: Implement Function
    #Not putting spaces in because this will be done by the helper.preprocess_and_save_data function
    tokens = {
        '.' : '<PERIOD>',
        ',' : '<COMMA>',
        '"' : '<QUOTATION_MARK>',
        ';' : '<SEMICOLON>',
        '!' : '<EXCLAMATION_MARK>',
        '?' : '<QUESTION_MARK>',
        '(' : '<LEFT_PAREN>',
        ')' : '<RIGHT_PAREN>',
        '-' : '<DASH>',
        '\n' : '<RETURN>'
    }
    
    return tokens

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_tokenize(token_lookup)

Tests Passed


## Pre-process all the data and save it

Running the code cell below will pre-process all the data and save it to file. You're encouraged to lok at the code for `preprocess_and_save_data` in the `helpers.py` file to see what it's doing in detail, but you do not need to change this code.

In [8]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# pre-process training data
helper.preprocess_and_save_data(data_dir, token_lookup, create_lookup_tables)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [9]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import udacity_helper as helper
import problem_unittests as tests

int_text, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()

## Build the Neural Network
In this section, you'll build the components necessary to build an RNN by implementing the RNN Module and forward and backpropagation functions.

### Check Access to GPU

In [10]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch

# Check for a GPU
train_on_gpu = torch.cuda.is_available()
if not train_on_gpu:
    print('No GPU found. Please use a GPU to train your neural network.')

## Input
Let's start with the preprocessed input data. We'll use [TensorDataset](http://pytorch.org/docs/master/data.html#torch.utils.data.TensorDataset) to provide a known format to our dataset; in combination with [DataLoader](http://pytorch.org/docs/master/data.html#torch.utils.data.DataLoader), it will handle batching, shuffling, and other dataset iteration functions.

You can create data with TensorDataset by passing in feature and target tensors. Then create a DataLoader as usual.
```
data = TensorDataset(feature_tensors, target_tensors)
data_loader = torch.utils.data.DataLoader(data, 
                                          batch_size=batch_size)
```

### Batching
Implement the `batch_data` function to batch `words` data into chunks of size `batch_size` using the `TensorDataset` and `DataLoader` classes.

>You can batch words using the DataLoader, but it will be up to you to create `feature_tensors` and `target_tensors` of the correct size and content for a given `sequence_length`.

For example, say we have these as input:
```
words = [1, 2, 3, 4, 5, 6, 7]
sequence_length = 4
```

Your first `feature_tensor` should contain the values:
```
[1, 2, 3, 4]
```
And the corresponding `target_tensor` should just be the next "word"/tokenized word value:
```
5
```
This should continue with the second `feature_tensor`, `target_tensor` being:
```
[2, 3, 4, 5]  # features
6             # target
```

In [11]:
from torch.utils.data import TensorDataset, DataLoader


def batch_data(words, sequence_length, batch_size):
    """
    Batch the neural network data using DataLoader
    :param words: The word ids of the TV scripts
    :param sequence_length: The sequence length of each batch
    :param batch_size: The size of each batch; the number of sequences in a batch
    :return: DataLoader with batched data
    """
    # DONE: Implement function

    features, targets = [],[]
    #Now loop through the data creating sequences. For each, create the features and target
    for idx in range(0, len(words)-sequence_length):
        
        sequence_features = words[idx:idx+sequence_length] #sequence containing the relevant words
        sequence_target = words[idx+sequence_length]
    
        features.append([feat for feat in sequence_features])
        targets.append([sequence_target])
    
    features = np.reshape(features, (len(features), sequence_length))
    targets = np.squeeze(targets)
    
    #finally create the Dataloader
    the_data = TensorDataset(torch.from_numpy(features), torch.from_numpy(targets))

    # make sure the SHUFFLE your training data
    the_loader = DataLoader(the_data, shuffle=False, batch_size=batch_size)
    
    # return a dataloader
    return the_loader

# there is no test for this function, but you are encouraged to create
# print statements and tests of your own


### Test your dataloader 

You'll have to modify this code to test a batching function, but it should look fairly similar.

Below, we're generating some test text data and defining a dataloader using the function you defined, above. Then, we are getting some sample batch of inputs `sample_x` and targets `sample_y` from our dataloader.

Your code should return something like the following (likely in a different order, if you shuffled your data):

```
torch.Size([10, 5])
tensor([[ 28,  29,  30,  31,  32],
        [ 21,  22,  23,  24,  25],
        [ 17,  18,  19,  20,  21],
        [ 34,  35,  36,  37,  38],
        [ 11,  12,  13,  14,  15],
        [ 23,  24,  25,  26,  27],
        [  6,   7,   8,   9,  10],
        [ 38,  39,  40,  41,  42],
        [ 25,  26,  27,  28,  29],
        [  7,   8,   9,  10,  11]])

torch.Size([10])
tensor([ 33,  26,  22,  39,  16,  28,  11,  43,  30,  12])
```

### Sizes
Your sample_x should be of size `(batch_size, sequence_length)` or (10, 5) in this case and sample_y should just have one dimension: batch_size (10). 

### Values

You should also notice that the targets, sample_y, are the *next* value in the ordered test_text data. So, for an input sequence `[ 28,  29,  30,  31,  32]` that ends with the value `32`, the corresponding output should be `33`.

In [12]:
# test dataloader

test_text = range(50)
t_loader = batch_data(test_text, sequence_length=5, batch_size=10)

data_iter = iter(t_loader)
sample_x, sample_y = data_iter.next()

print(sample_x.shape)
print(sample_x)
print()
print(sample_y.shape)
print(sample_y)

torch.Size([10, 5])
tensor([[ 0,  1,  2,  3,  4],
        [ 1,  2,  3,  4,  5],
        [ 2,  3,  4,  5,  6],
        [ 3,  4,  5,  6,  7],
        [ 4,  5,  6,  7,  8],
        [ 5,  6,  7,  8,  9],
        [ 6,  7,  8,  9, 10],
        [ 7,  8,  9, 10, 11],
        [ 8,  9, 10, 11, 12],
        [ 9, 10, 11, 12, 13]])

torch.Size([10])
tensor([ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14])


---
## Build the Neural Network
Implement an RNN using PyTorch's [Module class](http://pytorch.org/docs/master/nn.html#torch.nn.Module). You may choose to use a GRU or an LSTM. To complete the RNN, you'll have to implement the following functions for the class:
 - `__init__` - The initialize function. 
 - `init_hidden` - The initialization function for an LSTM/GRU hidden state
 - `forward` - Forward propagation function.
 
The initialize function should create the layers of the neural network and save them to the class. The forward propagation function will use these layers to run forward propagation and generate an output and a hidden state.

**The output of this model should be the *last* batch of word scores** after a complete sequence has been processed. That is, for each input sequence of words, we only want to output the word scores for a single, most likely, next word.

### Hints

1. Make sure to stack the outputs of the lstm to pass to your fully-connected layer, you can do this with `lstm_output = lstm_output.contiguous().view(-1, self.hidden_dim)`
2. You can get the last batch of word scores by shaping the output of the final, fully-connected layer like so:

```
# reshape into (batch_size, seq_length, output_size)
output = output.view(batch_size, -1, self.output_size)
# get last batch
out = output[:, -1]
```

In [13]:
import torch.nn as nn

class RNN(nn.Module):
    
    def __init__(self, vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5):
        """
        Initialize the PyTorch RNN Module
        :param vocab_size: The number of input dimensions of the neural network (the size of the vocabulary)
        :param output_size: The number of output dimensions of the neural network
        :param embedding_dim: The size of embeddings, should you choose to use them        
        :param hidden_dim: The size of the hidden layer outputs
        :param dropout: dropout to add in between LSTM/GRU layers
        """
        super(RNN, self).__init__()
        # DONE: Implement function
        
        # set class variables
        self.vocab_size = vocab_size
        self.output_size = output_size
        self.embedding_dim = embedding_dim
        self.hidden_dim = hidden_dim
        self.n_layers = n_layers
        self.dropout = nn.Dropout(dropout)
        
        # define model layers
        self.embedding = nn.Embedding(vocab_size, embedding_dim)
        
        
        self.lstm = nn.LSTM(input_size=embedding_dim, hidden_size=hidden_dim, num_layers=n_layers, 
                            dropout=dropout, batch_first=True)
        
        self.fc = nn.Linear(hidden_dim, output_size)

        #self.sm = nn.Softmax(dim=1)
    
    def forward(self, nn_input, hidden):
        """
        Forward propagation of the neural network
        :param nn_input: The input to the neural network
        :param hidden: The hidden state        
        :return: Two Tensors, the output of the neural network and the latest hidden state
        """
        # DONE: Implement function   
 
        #convert all contents of the tensor to integers
        #nn_input.long()
        embeds = self.embedding(nn_input)
        
        #now push through the ltsm layer
        lstm_out, hidden = self.lstm(embeds, hidden)
                
        #dropout and run through the fully connected layer
        out = self.dropout(lstm_out)
        out = self.fc(out)
        
        """
        Coming out of the fc layer, we'll have a shape of (batch_size, seq_inputs, outputs) 
        batch 0
            1st word in sequence 0
                output 1
                output 2
                output 3
                output ...
                output n
             2nd word in sequence 1
                output 1
                output 2
                output 3
                output ...
                output n    
            nth word in sequence - this is the set out outputs we're interested in for each batch
                output 1
                output 2
                output 3
                output ...
                output n               
         batch 1 ...
        """
        # but we only want the last set of outputs from each batch
        #probabilities = self.sm(out)
        #probabilities = probabilities[:,-1]
        #print (f"probabilities:{probabilities}")
        
        # Remove everything but the last sequence output
        probabilities2 = out[:,-1]
        #probabilities2 = self.sm(probabilities2)
        #print (f"probabilities2:{probabilities2}")
        
        # return one batch of output word scores and the hidden state
        return probabilities2, hidden
    
    
    def init_hidden(self, batch_size):
        '''
        Initialize the hidden state of an LSTM/GRU
        :param batch_size: The batch_size of the hidden state
        :return: hidden state of dims (n_layers, batch_size, hidden_dim)
        '''
        # Implement function
        
        # initialize hidden state with zero weights, and move to GPU if available
        weight = next(self.parameters()).data

        if (train_on_gpu):
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda(),
                    weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda())
        else:
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_(),
                        weight.new(self.n_layers, batch_size, self.hidden_dim).zero_())

        return hidden
   

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_rnn(RNN, train_on_gpu)

Tests Passed


### Define forward and backpropagation

Use the RNN class you implemented to apply forward and back propagation. This function will be called, iteratively, in the training loop as follows:
```
loss = forward_back_prop(decoder, decoder_optimizer, criterion, inp, target)
```

And it should return the average loss over a batch and the hidden state returned by a call to `RNN(inp, hidden)`. Recall that you can get this loss by computing it, as usual, and calling `loss.item()`.

**If a GPU is available, you should move your data to that GPU device, here.**

In [14]:
def forward_back_prop(rnn, optimizer, criterion, inp, target, hidden):
    """
    Forward and backward propagation on the neural network
    :param rnn: The PyTorch Module that holds the neural network
    :param optimizer: The PyTorch optimizer for the neural network
    :param criterion: The PyTorch loss function
    :param inp: A batch of input to the neural network
    :param target: The target output for the batch of input
    :return: The loss and the latest hidden state Tensor
    """
    
    # DONE: Implement Function
    #Put the model into training mode
    rnn.train()  

    # move data to GPU, if available
    if (train_on_gpu):
        rnn.cuda()
        inp, target = inp.cuda(), target.cuda() 

    # perform backpropagation and optimization
    # Creating new variables for the hidden state, otherwise
    # we'd backprop through the entire training history
    hidden = tuple([each.data for each in hidden]) 

    # zero accumulated gradients
    rnn.zero_grad()
    
    # get the output from the model
    output, hidden = rnn(inp, hidden)
    
    #Output returns a set of probabilities so convert to each output to a single value based
    # on the index number of the most probable
    #print (f"target:{target.shape} {target}")
    #print (f"output:{output.shape} {output}")   
    
    #predictions = torch.argmax(output, dim=1)   
    #print (f"prediction:{predictions.shape} {predictions}")
    
    # calculate the loss and perform backprop
    loss = criterion(output, target)
    loss.backward()
    # `clip_grad_norm` helps prevent the exploding gradient problem in RNNs / LSTMs.
    nn.utils.clip_grad_norm_(rnn.parameters(),5)
    optimizer.step() 

    # return the loss over a batch and the hidden state produced by our model
    return loss.item(), hidden

# Note that these tests aren't completely extensive.
# they are here to act as general checks on the expected outputs of your functions
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_forward_back_prop(RNN, forward_back_prop, train_on_gpu)

Tests Passed


## Neural Network Training

With the structure of the network complete and data ready to be fed in the neural network, it's time to train it.

### Train Loop

The training loop is implemented for you in the `train_decoder` function. This function will train the network over all the batches for the number of epochs given. The model progress will be shown every number of batches. This number is set with the `show_every_n_batches` parameter. You'll set this parameter along with other parameters in the next section.

In [15]:
from datetime import datetime

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

def train_rnn(rnn, batch_size, optimizer, criterion, n_epochs, show_every_n_batches=100):
    batch_losses = []
    
    rnn.train()

    now = datetime.now()
    current_time = now.strftime("%H:%M:%S")
    print(f"{current_time} Training for {n_epochs:d} epoch(s)...")
    for epoch_i in range(1, n_epochs + 1):
        
        # initialize hidden state
        hidden = rnn.init_hidden(batch_size)
        
        for batch_i, (inputs, labels) in enumerate(train_loader, 1):
            
            # make sure you iterate over completely full batches, only
            n_batches = len(train_loader.dataset)//batch_size
            if(batch_i > n_batches):
                break
            
            # forward, back prop
            loss, hidden = forward_back_prop(rnn, optimizer, criterion, inputs, labels, hidden)          
            # record loss
            batch_losses.append(loss)

            # printing loss stats
            now = datetime.now()
            current_time = now.strftime("%H:%M:%S")
            if batch_i % show_every_n_batches == 0:
                print('{} Epoch: {:>4}/{:<4}  Loss: {}\n'.format(
                    current_time, epoch_i, n_epochs, np.average(batch_losses)))
                batch_losses = []

    # returns a trained rnn
    return rnn

### Hyperparameters

Set and train the neural network with the following parameters:
- Set `sequence_length` to the length of a sequence.
- Set `batch_size` to the batch size.
- Set `num_epochs` to the number of epochs to train for.
- Set `learning_rate` to the learning rate for an Adam optimizer.
- Set `vocab_size` to the number of uniqe tokens in our vocabulary.
- Set `output_size` to the desired size of the output.
- Set `embedding_dim` to the embedding dimension; smaller than the vocab_size.
- Set `hidden_dim` to the hidden dimension of your RNN.
- Set `n_layers` to the number of layers/cells in your RNN.
- Set `show_every_n_batches` to the number of batches at which the neural network should print progress.

If the network isn't getting the desired results, tweak these parameters and/or the layers in the `RNN` class.

In [16]:
# Data params
# Sequence Length
sequence_length =  15 # of words in a sequence
# Batch Size
batch_size = 128

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

In [28]:
# Training parameters
# Number of Epochs
num_epochs = 5
# Learning Rate
learning_rate = 0.0001

# Model parameters
# Vocab size
vocab_size = len(int_to_vocab)
# Output size
output_size = vocab_size
# Embedding Dimension
embedding_dim = 500
# Hidden Dimension
hidden_dim = 500
# Number of RNN Layers
n_layers = 3

# Show stats for every n number of batches
show_every_n_batches = 2000

### Train
In the next cell, you'll train the neural network on the pre-processed data.  If you have a hard time getting a good loss, you may consider changing your hyperparameters. In general, you may get better results with larger hidden and n_layer dimensions, but larger models take a longer time to train. 
> **You should aim for a loss less than 3.5.** 

You should also experiment with different sequence lengths, which determine the size of the long range dependencies that a model can learn.

In [18]:
rnn = helper.load_model('./save/trained_rnn')

In [29]:
#import workspace_utils
#from workspace_utils import active_session
 
  #with active_session():

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

# create model and move to gpu if available
rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
if train_on_gpu:
    rnn.cuda()

# defining loss and optimization functions for training
optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
criterion = nn.CrossEntropyLoss()

# training the model
trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/trained_rnn', trained_rnn)
print('Model Trained and Saved')

20:42:07 Training for 5 epoch(s)...
20:43:31 Epoch:    1/5     Loss: 3.487785143494606

20:44:55 Epoch:    1/5     Loss: 3.530271303653717

20:46:20 Epoch:    1/5     Loss: 3.514893903851509

20:48:25 Epoch:    2/5     Loss: 3.5165670015226516

20:49:49 Epoch:    2/5     Loss: 3.525316744327545

20:51:14 Epoch:    2/5     Loss: 3.506552477478981

20:53:19 Epoch:    3/5     Loss: 3.5099935082724376

20:54:43 Epoch:    3/5     Loss: 3.517904278755188

20:56:08 Epoch:    3/5     Loss: 3.50187160384655

20:58:13 Epoch:    4/5     Loss: 3.5016029621463467

20:59:37 Epoch:    4/5     Loss: 3.511075606942177

21:01:01 Epoch:    4/5     Loss: 3.4983957438468933

21:03:07 Epoch:    5/5     Loss: 3.498471126366882

21:04:31 Epoch:    5/5     Loss: 3.505381284236908

21:05:55 Epoch:    5/5     Loss: 3.494655014038086

Model Trained and Saved


### Question: How did you decide on your model hyperparameters? 
For example, did you try different sequence_lengths and find that one size made the model converge faster? What about your hidden_dim and n_layers; how did you decide on those?

**Answer:** (Write answer, here)

__Summary__

|Attempt| sequence length 	| batch size 	| hidden dim 	| n_layers 	|      embed dim     	| lr   	| loss  	|
|-|:---------------:	|:----------:	|:----------:	|:--------:	|:------------------:	|------	|:----------------------:	|
|1| 50              	| 50         	| 200        	| 10       	| 1/2 vocab size     	| 0.01 	| 7 batches: 9.97124     	|
|2| 1000            	| 100        	| 200        	| 10       	| 1/2 vocab size     	| 0.01 	| out of memory error    	|
|3| 500            	| 20         	| 200        	| 10       	| 1/2 vocab size    	| 0.01 	| 2 batches: 9.970524038                       	|
|4| 500            	| 20         	| 200        	| 2       	| 100    	| 0.001 |  12 mins: 9.9654329                       	|
|5| 500            	| 20         	| 200        	| 2       	| 100    	| 0.01 |  15 mins: 9.970885                      	|
|6| 50            	| 32         	| 200        	| 2       	| 100    	| 0.01 |  35 mins: 9.958274                      	|
|7| 50            	| 32         	| 200        	| 2       	| 500    	| 0.01 |  38 mins: 9.942556                      	|
|8| 50            	| 128         	| 256        	| 2       	| 256    	| 0.01 |                        	|
|9| 50            	| 128         	| 256        	| 2       	| 256    	| 0.1 | 52 mins : 5.4708849                       	|
|10| 50            	| 128         	| 256        	| 2       	| 256    	| 0.1 | Continued training from attempt 9: 5.938092|
|11| 50            	| 128         	| 256        	| 2       	| 256    	| 0.01 | 18 mins: 4.9209547|
|12| 50            	| 128         	| 256        	| 2       	| 256    	| 0.01 | Stopped at 9 epochs. loss levelled out|
|13| 5            	| 128         	| 256        	| 3       	| 256    	| 0.1 |Loss stalled at around 5.7885441 |
|14| 5            	| 128         	| 256        	| 3       	| 256    	| 0.01 |Loss stalled at around 5.6 |
|15| 5            	| 128         	| 256        	| 3       	| 500    	| 0.001 | 11 mins: 4.08200105|
|16| 15            	| 128         	| 256        	| 3       	| 500    	| 0.001 | 1.5hrs: 3.7618935176849364 |
|17| 15            	| 128         	| 350        	| 3       	| 500    	| 0.001 | 2hrs: 3.69475595 |
|18| 15            	| 128         	| 350        	| 3       	| 500    	| 0.001 | 2hrs: 3.6054698 |
|19| 15            	| 128         	| 500        	| 3       	| 500    	| 0.001 | 2hrs: 3.5265116 |
|20| 15            	| 128         	| 500        	| 3       	| 500    	| 0.0001 | 3.49465501 |

__Attempt 1:__
Sequence length - Started with a 50. Whilst the average number of words in a line was around 5, I wanted to capture a sequnce that encompassed a number of lines at once so I chose 50
Batch size - started with a fairly small batch size so I could see results a bit quicker
hidden state - quite a large vocabularly so chose quite a lot of features

__Attempt 4:__
With no real improvement in loss, I changed the hyperparameters based on research papers from the RMM Hyperparapmeters lesson. I also lowered the learning rate in case it was jumping too far each time. Performance was better, but was improving very slowly.

__Attempt 5:__
Increased the learning rate.

__Attempt 6:__
Changed sequence length to 50 as per attempt 1, so now the hidden dim is bigger than the input. Increased batch size to 32 to improve training time. Accuracy, whilst better than other models, did not show a great improvement through batches in the epoch

__Attempt 7:__
Increased embedding dimensions to 500 to see whether that increases accuracy. 

__Attempt 8:__
No real improvement in attempt 7. Checked model for this attemp and removed the final softmax layer along with increasing batch size and changing hidden state and embedded dim to be the same. Achieved steady reduction in loss, but still slow

__Attempt 9:__
Increased the learning rate to 0.1. Training appears successful. Continuing training on using this model for more epochs.

__Attempt 10__:
Continued from attempt 9. Did not improve. Enough combinations of hyperparameters have taken place without any great improvement. Need to revisit the model to check its right.

__Attempt 11__:
Something feels wrong about removing the softmax layer as we should be dealing with probability of a particular word being the next in the sequence. Turns out that the torch.CrossEntropyLoss function is just expecting raw outputs as it will do the softmax itself.

__Attempt 12:__
Continued training from attempt 11. Workspace timed out and I lost the training that had been done. Introduced a workspace utility to keep the workspace alive. On restart the loss levelled out at around 5.25 after 9 epochs

__Attempt 13__
Increased layers to 3 and reduced learning sequence length to 5 see if it could achieve better accuracy. TRaining still hit a plateau of around 5.7, even after 50 epochs

__Attempt_14:__
Turned off shuffling in the batches and decreased learning rate to 0.01 to see if it can get out of the plateau. This should be ok, as I actually want the model to take into account the order in which lines occur. Running 2 epochs at first

__Attempt_15:__
Increased embedding layer size as the model probably hasn't got enough capacity to detect patterns. Run for 10 epochs. Much more success with this. This is almost the same hyperparameters as attempt 7 but with smaller sequence length, more batches and an extra layer. Continually improved over 10 epochs

__Attempt_16:__
Although I could probably reach the required accuracy with the model from attempt 15, I increased the sequence length to take into account more history. Hopefully this will give me a better model which has better output

__Attempt_17:__
Continued training Attempt 16 for another 20 epochs

__Attempt_18:__
Continued training Attempt 18 for another 20 epochs. Looks like this model has reached as fa as it will go. Hasn't quite got to 3.5 so need to revisit the hyperparameters

__Attempt_19:__
Increased the hidden layer to see if it can detect more patterns

__Attempt_20:__
Achieved 3.52 but appeared to stall after 40 epochs. Dropped learning rate to 0.0001 for next 5 epochs and achieved sub 3.5 loss


---
# Checkpoint

After running the above training cell, your model will be saved by name, `trained_rnn`, and if you save your notebook progress, **you can pause here and come back to this code at another time**. You can resume your progress by running the next cell, which will load in our word:id dictionaries _and_ load in your saved model by name!

In [30]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch
import udacity_helper as helper
import problem_unittests as tests

_, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()
trained_rnn = helper.load_model('./save/trained_rnn')

## Generate TV Script
With the network trained and saved, you'll use it to generate a new, "fake" Seinfeld TV script in this section.

### Generate Text
To generate the text, the network needs to start with a single word and repeat its predictions until it reaches a set length. You'll be using the `generate` function to do this. It takes a word id to start with, `prime_id`, and generates a set length of text, `predict_len`. Also note that it uses topk sampling to introduce some randomness in choosing the most likely next word, given an output set of word scores!

In [31]:
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import torch.nn.functional as F

def generate(rnn, prime_id, int_to_vocab, token_dict, pad_value, predict_len=100):
    """
    Generate text using the neural network
    :param decoder: The PyTorch Module that holds the trained neural network
    :param prime_id: The word id to start the first prediction
    :param int_to_vocab: Dict of word id keys to word values
    :param token_dict: Dict of puncuation tokens keys to puncuation values
    :param pad_value: The value used to pad a sequence
    :param predict_len: The length of text to generate
    :return: The generated text
    """
    rnn.eval()
    
    # create a sequence (batch_size=1) with the prime_id
    current_seq = np.full((1, sequence_length), pad_value)
    current_seq[-1][-1] = prime_id
    predicted = [int_to_vocab[prime_id]]
    
    for _ in range(predict_len):
        if train_on_gpu:
            current_seq = torch.LongTensor(current_seq).cuda()
        else:
            current_seq = torch.LongTensor(current_seq)
        
        # initialize the hidden state
        hidden = rnn.init_hidden(current_seq.size(0))
        
        # get the output of the rnn
        output, _ = rnn(current_seq, hidden)
        
        # get the next word probabilities
        p = F.softmax(output, dim=1).data
        if(train_on_gpu):
            p = p.cpu() # move to cpu
         
        # use top_k sampling to get the index of the next word
        top_k = 5
        p, top_i = p.topk(top_k)
        top_i = top_i.numpy().squeeze()
        
        # select the likely next word index with some element of randomness
        p = p.numpy().squeeze()
        word_i = np.random.choice(top_i, p=p/p.sum())
        
        # retrieve that word from the dictionary
        word = int_to_vocab[word_i]
        predicted.append(word)     
        
        # the generated word becomes the next "current sequence" and the cycle can continue
        current_seq = np.roll(current_seq.cpu(), -1, 1)
        current_seq[-1][-1] = word_i
    
    gen_sentences = ' '.join(predicted)
    
    # Replace punctuation tokens
    for key, token in token_dict.items():
        ending = ' ' if key in ['\n', '(', '"'] else ''
        gen_sentences = gen_sentences.replace(' ' + token.lower(), key)
    gen_sentences = gen_sentences.replace('\n ', '\n')
    gen_sentences = gen_sentences.replace('( ', '(')
    
    # return all the sentences
    return gen_sentences

### Generate a New Script
It's time to generate the text. Set `gen_length` to the length of TV script you want to generate and set `prime_word` to one of the following to start the prediction:
- "jerry"
- "elaine"
- "george"
- "kramer"

You can set the prime word to _any word_ in our dictionary, but it's best to start with a name for generating a TV script. (You can also start with any other names you find in the original text file!)

In [32]:
# run the cell multiple times to get different results!
gen_length = 400 # modify the length to your preference
prime_word = 'jerry' # name for starting the script

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
pad_word = helper.SPECIAL_WORDS['PADDING']
generated_script = generate(trained_rnn, vocab_to_int[prime_word + ':'], int_to_vocab, token_dict, vocab_to_int[pad_word], gen_length)
print(generated_script)

jerry:" ]

kramer:(laughing to elaine) well, i guess i could.

george:(getting up to leave) well, i'm sorry, i'm gonna have to get a piece of time.

elaine: oh, i don't know how you feel that.

george: you know, i have a very little thing with you, huh?

george:(sarcastic) oh, no- no, that's not.

elaine:(looking at the window) hey, hey, hey! hey!

kramer: hey, hey!

jerry: hey, hey, hey!

newman: jerry!

newman: hi, jerry.

elaine: hi.

elaine: what are you talking about?

jerry: well, i don't have any money.

elaine: no.

jerry: so you can get a little job?

george: no, you don't know.

elaine: i think i should have had the money.

jerry: oh, you don't know, the new yorker.

kramer: yeah, i think we could.

elaine: you know, it's a little bit of the big time.

jerry:(confused) well, it's not the most thing.

elaine: well, i can't believe you don't have a good time.

kramer: oh, no. i don't think we have.

jerry: you know, i have to go back to the movies and i can go.

jerry: so, you 

#### Save your favorite scripts

Once you have a script that you like (or find interesting), save it to a text file!

In [33]:
# save script to a text file
f =  open("generated_script_1.txt","w")
f.write(generated_script)
f.close()

# The TV Script is Not Perfect
It's ok if the TV script doesn't make perfect sense. It should look like alternating lines of dialogue, here is one such example of a few generated lines.

### Example generated script

>jerry: what about me?
>
>jerry: i don't have to wait.
>
>kramer:(to the sales table)
>
>elaine:(to jerry) hey, look at this, i'm a good doctor.
>
>newman:(to elaine) you think i have no idea of this...
>
>elaine: oh, you better take the phone, and he was a little nervous.
>
>kramer:(to the phone) hey, hey, jerry, i don't want to be a little bit.(to kramer and jerry) you can't.
>
>jerry: oh, yeah. i don't even know, i know.
>
>jerry:(to the phone) oh, i know.
>
>kramer:(laughing) you know...(to jerry) you don't know.

You can see that there are multiple characters that say (somewhat) complete sentences, but it doesn't have to be perfect! It takes quite a while to get good results, and often, you'll have to use a smaller vocabulary (and discard uncommon words), or get more data.  The Seinfeld dataset is about 3.4 MB, which is big enough for our purposes; for script generation you'll want more than 1 MB of text, generally. 

# Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_tv_script_generation.ipynb" and save another copy as an HTML file by clicking "File" -> "Download as.."->"html". Include the "helper.py" and "problem_unittests.py" files in your submission. Once you download these files, compress them into one zip file for submission.